### PR TITLE
[GraphBolt][CUDA] Specialize non-weighted neighbor sampling impl

### LINKED ISSUE #7173: [GraphBolt][CUDA] Specialized kernels for simple Neighbor Sampling

### PR DESCRIPTION
**Pull request by mfbalin**
Fixes #7173. Neighbor Sampling kernel is sped up by 4-5x compared to old, non-specialized version.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

**Comment by mfbalin** [comment]
@TristonC The CI has the following error message:
```
/home/ubuntu/jenkins/workspace/dgl_PR-7215/third_party/cccl/libcudacxx/include/cuda/std/detail/libcxx/include/support/atomic/atomic_cuda.h:12:4: error: #error "CUDA atomics are only supported for sm_60 and up on *nix and sm_70 and up on Windows."
   12 | #  error "CUDA atomics are only supported for sm_60 and up on *nix and sm_70 and up on Windows."
```

File: graphbolt/src/cuda/neighbor_sampler.cu [hunk 1]
```diff
@@ -30,6 +33,64 @@ namespace ops {
 
 constexpr int BLOCK_SIZE = 128;
 
+inline __device__ int64_t AtomicMax(int64_t* const address, const int64_t val) {
+  // match the type of "::atomicCAS", so ignore lint warning
```
Comment (frozenbugs) [inline graphbolt/src/cuda/neighbor_sampler.cu:37]:
```suggestion
  // To match the type of "::atomicCAS", ignore lint warning.
```

File: graphbolt/src/cuda/neighbor_sampler.cu [hunk 2]
```diff
@@ -30,6 +33,64 @@ namespace ops {
...
+  using Type = unsigned long long int;  // NOLINT
+
+  static_assert(sizeof(Type) == sizeof(*address), "Type width must match");
+
+  return atomicMax(reinterpret_cast<Type*>(address), static_cast<Type>(val));
+}
+
+inline __device__ int32_t AtomicMax(int32_t* const address, const int32_t val) {
+  // match the type of "::atomicCAS", so ignore lint warning
```
Comment (frozenbugs) [inline graphbolt/src/cuda/neighbor_sampler.cu:46]:
```suggestion
  // To match the type of "::atomicCAS", ignore lint warning.
```

File: graphbolt/src/cuda/neighbor_sampler.cu [hunk 3]
```diff
@@ -251,119 +312,186 @@ c10::intrusive_ptr<sampling::FusedSampledSubgraph> SampleNeighbors(
...
+              torch::kLong};
+          auto edge_id_dtype = types[type_index];
+          AT_DISPATCH_INTEGRAL_TYPES(
+              edge_id_dtype, "SampleNeighborsEdgeIDs", ([&] {
+                using edge_id_t = std::make_unsigned_t<scalar_t>;
+                TORCH_CHECK(
+                    num_bits <= sizeof(edge_id_t) * 8,
+                    "Selected edge_id_t must be capable of storing edge_ids.");
+                // Using bfloat16 for random numbers works just as reliably as
+                // float32 and provides around %30 percent speedup.
```
Comment (frozenbugs) [inline graphbolt/src/cuda/neighbor_sampler.cu:336]:
```suggestion
                // float32 and provides around 30% speedup.
```

**Review by frozenbugs — APPROVED** [review]

**Pull request merged**